### ISSUE TITLE
Keyed Services Support? 🤔

### ISSUE THREAD

**Issue by Mike-E-angelo** [seesharper/LightInject#596]
This issue piggybacks on this one:
https://github.com/seesharper/LightInject.Microsoft.DependencyInjection/issues/210

I am beginning to wonder... is this a bigger problem?  Does LightInject need to support Keyed Services? 🤔

**Comment by seesharper** [comment]
Hi @Mike-E-angelo 

The problem here is a little bit more complicated than it might seem at first glance.
LightInject has always supported keyed/named services so one should think that this should work out of the box.
The problem is that `LightInject.Microsoft.DependencyInjection` does some , should we call it "dirty tricks" to make LightInject compatible with MS.DI. 😃 Some of those tricks involves an internal naming of services so that we for instance can bring back services in their registration order. 
Take a look here 
https://github.com/seesharper/LightInject.Microsoft.DependencyInjection/blob/9e5424007878ff0cf4b54ab9c0ce9402b4a36bc6/src/LightInject.Microsoft.DependencyInjection/LightInject.Microsoft.DependencyInjection.cs#L118

It ain't pretty,but it has worked up until now where MS.DI itself has started to support keyed services. Then those "hacks" would no longer work in all cases and the adapter breaks down. 

So to fix this I've started to do some changes to LightInject itself to better adapt to MS.DI without using LightInject's named services as a way to be compatible with MS.DI. 

I've started this work some time ago, but since nobody has asked for it yet , I pretty much left it on the shelf knowing that I would have to pick up on it eventually. 

I guess that time is now. But just as you know, there is no easy fix here that can be solved in the adapter. Changes has to be made in LightInject to accommodate this.

**Comment by seesharper** [comment]
We probably end up with a new option in LightInject.
Something like `ContainerOptions.EnableMicrosoftCompatibility` or something like that. 
Once that is in place and it passes all the tests from MS.DI with the option enabled it should be fairly easy to modify the adapter. In fact the adapter would be much simpler since we don't need to "adapt" so much anymore. 😃

**Comment by Mike-E-angelo** [comment]
That sounds awesome @seesharper thank you for taking the time to comment and inform.  I have to admit I am doing double-takes here because I am now shaking my head that "hey! DI has always had named services." ... 🤔  So these keyed services are different?  I am so confused. 😅

I am of course more than happy to help test any new features that help address this problem.  As I mentioned in the other thread an error is thrown currently but soon it will be null and that might lead to other issues.

**Comment by seesharper** [comment]
Hi Mike. Just wanted to let you know that I have been working on this quite a lot the last couple of weeks. I have all the spec tests from MS working now. Still some cleanup needed, but we're getting there. Should this ready by the time .net 9 goes RTM 👍🙂

**Comment by Mike-E-angelo** [comment]
That's awesome @seesharper!  Thank you so very much for all your efforts out there. 🙏

**Comment by seesharper** [comment]
Thanks a lot for sponsoring me. 🤩. Really appreciate it.

**Comment by Mike-E-angelo** [comment]
Right back atcha @seesharper ... thank you for being so responsive and engaging with your product.  It has been fundamental and utilized by my framework for many years and is the least I can do. 🙏

**Comment by Mike-E-angelo** [comment]
Awesome @seesharper thank you very much for the update.  For your awareness, the project I was working that was using Aspire (where I ran into this issue) has been put on hold for now, so this is not a rush/priority at the moment, but it would of course be great to have it there when I need it. 😊🙏  I appreciate all your work out there and wish a great weekend for you as well. 👍

**Comment by seesharper** [comment]
Hi @Mike-E-angelo 
Just wanted to inform you that support for keyed services is now fully implemented.

The new packages are 
* LightInject 7.0.0
* LightInject.Microsoft.DependencyInjection 4.0.0
* LightInject.Microsoft.Hosting 2.0.0

Technically there are no breaking changes in these versions, but I decided to bump the major anyway 🙂

Again thanks for your sponsor contribution and for the encouragement ❤️

**Comment by Mike-E-angelo** [comment]
That's really awesome @seesharper!  Thank you again so very much for this.  This will be very useful going forward to onboard the new tech such as Aspire which is using this feature. 🙏  I greatly appreciate all your efforts out there. 👍✨

**Comment by Mike-E-angelo** [comment]
Hi @seesharper I hope this finds you well out there.  I have run into an unexpected error and I wanted to be sure I understand the required implementation.  Using the following:

```
<PackageVersion Include="LightInject.Microsoft.AspNetCore.Hosting" Version="2.3.5" />
<PackageVersion Include="LightInject.Microsoft.Hosting" Version="2.2.0" />
<PackageVersion Include="Microsoft.Extensions.Caching.Hybrid" Version="10.9.0" />
```

Calling [this](https://github.com/DragonSpark/Framework/blob/release/v2/DragonSpark.Application.Hosting.Server.Blazor/DistributedAwareServiceConfiguration.cs#L22):

```
public void Execute(IServiceCollection parameter)
{
	parameter.AddHybridCache();
	parameter.AddOptions<CircuitOptions>()
	         .Configure<HybridCache>((options, subject) => options.HybridPersistenceCache = subject);
	_previous.Execute(parameter);
}
```
I get an error:

```
System.InvalidOperationException: A problem was encountered while resolving type Microsoft.AspNetCore.Authorization.Policy.AuthorizationPolicyCache
 ---> System.InvalidOperationException: A problem was encountered while resolving type Microsoft.AspNetCore.Components.Server.Circuits.CircuitRegistry
 ---> System.InvalidOperationException: A problem was encountered while resolving type System.Collections.Generic.IEnumerable`1[Microsoft.Extensions.Caching.Hybrid.IHybridCacheSerializerFactory]
 ---> System.InvalidOperationException: This service provider doesn't support keyed services.
   at object DynamicMethod(object[], Scope) x 2
   at object LightInject.ServiceContainer.TryGetInstance(Type serviceType, Scope scope)
   at object LightInject.Scope.TryGetInstance(Type serviceType)
   at object LightInject.Microsoft.DependencyInjection.LightInjectServiceProvider.GetService(Type serviceType)
   at object DragonSpark.Composition.ActivationAwareServiceProvider.GetService(Type serviceType) in .../Framework/DragonSpark.Composition/ActivationAwareServiceProvider.cs:line 33
   --- End of inner exception stack trace ---
   at object DragonSpark.Composition.ActivationAwareServiceProvider.GetService(Type serviceType) in .../Framework/DragonSpark.Composition/ActivationAwareServiceProvider.cs:line 41
   at object Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)
   at T Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService<T>(IServiceProvider provider)
   at new Microsoft.Extensions.Caching.Hybrid.Internal.DefaultHybridCache(HybridCacheOptions options, IServiceProvider services)
   at object DynamicMethod(object[], Scope) x 2
   at object LightInject.Scope.GetInstance(Type serviceType)
   at object LightInject.Microsoft.DependencyInjection.LightInjectServiceProvider.GetRequiredService(Type serviceType)
   at T Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService<T>(IServiceProvider provider)
   at OptionsBuilder<TOptions> Microsoft.Extensions.Options.OptionsBuilder<TOptions>.Configure<TDep>(Action<TOptions, TDep> configureOptions)+(IServiceProvider sp) => { }
   at Func<IServiceFactory, T> LightInject.Microsoft.DependencyInjection.DependencyInjectionContainerExtensions.CreateTypedFactoryDelegate<T>(ServiceDescriptor serviceDescriptor)+(IServiceFactory serviceFactory) => { }
   at object DynamicMethod(object[], Scope) x 3
   at object LightInject.ServiceContainer.TryGetInstance(Type serviceType, Scope scope)
   at object LightInject.Scope.TryGetInstance(Type serviceType)
   at object LightInject.Microsoft.DependencyInjection.LightInjectServiceProvider.GetService(Type serviceType)
   at object DragonSpark.Composition.ActivationAwareServiceProvider.GetService(Type serviceType) in .../Framework/DragonSpark.Composition/ActivationAwareServiceProvider.cs:line 33
```

Please let me know if another configuration or more information is necessary.  Thank you again for all your efforts there. 🙏

**Comment by seesharper** [comment]
Hi @Mike-E-angelo . The link that you provided did not work. I guess the pr was merged and/or the branch was deleted.  Could you point me to where it goes south for you?🙂

**Comment by Mike-E-angelo** [comment]
Hi @seesharper, never mind this. 🤦  User error.  LightInject implements IKeyedServiceProvider, but I do not pass it along with my own implementation/wrapper.  I will resolve this on my end.

I will also close this issue as it seems to be fixed/addressed.  Thank you again for all your work and effort, and for your patience with those of us who are a little too quick to jump on GitHub issues (even if it's a good chance to check in and say hi!).

**Comment by seesharper** [comment]
No worries 👍❤️
